platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/fingerprint/ConfigurationCacheFingerprintChecker.kt

```diff
@@ -348,14 +358,29 @@ class ConfigurationCacheFingerprintChecker(private val host: Host) {
         host.displayNameOf(file)
 
     private
-    fun buildLogicInputHasChanged(valueSource: ValueSource<Any, ValueSourceParameters>): InvalidationReason =
+    fun buildLogicInputHasChanged(valueSource: ValueSource<Any, ValueSourceParameters>): InvalidationReason = StructuredMessage.forText(
         (valueSource as? Describable)?.let {
             it.displayName + " has changed"
         } ?: "a build logic input of type '${unpackType(valueSource).simpleName}' has changed"
+    )
```

Comment:
In the current implementation, some `Describable`s can return strings like `system property 'BLAH'`. Since they are unstructured strings, we keep them as such for now. But this can be improved in the future by spreading the concept of structured messages wider in the code. For instance, we could allow `Describable`s to optionally provide a structured message as well.